lsp-mode.el

```diff
@@ -4429,44 +4429,48 @@ Also, additional data to attached to each candidate can be passed via PLIST."
        :exit-function
-       (lambda (candidate _status)
-         (-let* (((&plist 'lsp-completion-item item
-                          'lsp-completion-start-point start-point
-                          'lsp-completion-prefix-line prefix-line)
-                  (text-properties-at 0 candidate))
-                 ((&hash "label"
-                         "insertText" insert-text
-                         "textEdit" text-edit
-                         "insertTextFormat" insert-text-format
-                         "additionalTextEdits" additional-text-edits)
-                  item))
-           (cond
-            (text-edit
-             (delete-region (point-at-bol) (point))
-             (insert prefix-line)
-             (lsp--apply-text-edit text-edit))
-            ((or insert-text label)
-             (delete-region (point-at-bol) (point))
-             (insert prefix-line)
-             (delete-region start-point (point))
-             (insert (or insert-text label))))
-
-           (when (eq insert-text-format 2)
-             (yas-expand-snippet
-              (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
-              start-point
-              (point)))
-           (when (and lsp-completion-enable-additional-text-edit additional-text-edits)
-             (lsp--apply-text-edits additional-text-edits)))
-         (lsp--capf-clear-cache)
-         (when (and lsp-signature-auto-activate
-                    (lsp-feature? "textDocument/signatureHelp"))
-           (lsp-signature-activate))
-
-         (setq-local lsp-inhibit-lsp-hooks nil)
-
-         (when (lsp--looking-back-trigger-characterp trigger-chars)
-           (setq this-command 'self-insert-command)))))))
+       (-rpartial #'lsp--capf-exit-fn trigger-chars)))))
+
+(defun lsp--capf-exit-fn (candidate _status &optional trigger-chars)
+  "Exit function of `completion-at-point'.
+CANDIDATE is the selected completion item.
+Others: TRIGGER-CHARS"
+  (-let* (((&plist 'lsp-completion-item item
+                   'lsp-completion-start-point start-point
+                   'lsp-completion-markers markers)
+           (text-properties-at 0 candidate))
+          ((&hash "label"
+                  "insertText" insert-text
+                  "textEdit" text-edit
+                  "insertTextFormat" insert-text-format
+                  "additionalTextEdits" additional-text-edits)
+           item))
+    (let ((bol (point-at-bol)))
+      (cond
+       (text-edit
+        (apply #'delete-region markers)
+        (lsp--apply-text-edit text-edit))
+       ((or insert-text label)
+        (apply #'delete-region markers)
+        (delete-region start-point (point))
+        (insert (or insert-text label)))))
+
+    (when (eq insert-text-format 2)
+      (yas-expand-snippet
+       (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
+       start-point
+       (point)))
+    (when (and lsp-completion-enable-additional-text-edit additional-text-edits)
```

Comment:
I wonder whether we could move this before the yas-expand which will solve the issue with overlays protection?

Reply:
Read the code for overlays protection again, I think additional text edits will not affect to yas's overlays. Since according to the specs, the additional text edits cannot interferes with the main edit. Thus even modify them, I think the overlay protection will not be triggered.